### PR TITLE
gpt2 contiguous fix

### PR DESCRIPTION
**Pull request by BaihuiJin**
# What does this PR do?

change tensor to contiguous after split heads and before attention for gpt2

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by BaihuiJin** [comment]
With generation task, max_new_token=100 and bs=100
Before: 
Throughput (including tokenization) = 13081.375516379572 tokens/second
Memory allocated                    = 1.94 GB
Max memory allocated                = 28.02 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 3.4164320450508967 seconds
After:
Throughput (including tokenization) = 25008.293587717493 tokens/second
Memory allocated                    = 2.09 GB
Max memory allocated                = 3.48 GB
Total memory available              = 94.46 GB
Graph compilation duration          = 2.0869328420376405 seconds

**Comment by BaihuiJin** [comment]
image
After these changes:
image

**Review by regisss — COMMENTED** [review]
LGTM!

The code style check failed, could you run the following from the root of the repo to make sure it passes please?
```
pip install --upgrade black ruff
make style
```

**Comment by BaihuiJin** [comment]
> LGTM!
> 
> The code style check failed, could you run the following from the root of the repo to make sure it passes please?
> 
> ```
> pip install --upgrade black ruff
> make style
> ```

Got it, thanks!

**Review by regisss — APPROVED** [review]
I'll let the final approval to @ZhaiFeiyue

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 1]
```diff
@@ -75,6 +75,8 @@ def prune_heads(self, heads):
         self.pruned_heads = self.pruned_heads.union(heads)
 
     def _attn(self, query, key, value, attention_mask=None, head_mask=None):
+        key = key.contiguous()
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/gpt2/modeling_gpt2.py:78]: with line 212, do we still need contiguous here?

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py [hunk 2]
```diff
@@ -115,6 +117,8 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
         return attn_output, attn_weights
 
     def _upcast_and_reordered_attn(self, query, key, value, attention_mask=None, head_mask=None):
+        key = key.contiguous()
```
Comment (ZhaiFeiyue) [inline optimum/habana/transformers/models/gpt2/modeling_gpt2.py:120]: here you tested this branch?

**Comment by ZhaiFeiyue** [comment]
@BaihuiJin comments added, please have a look

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py (hunk 1 above)
Comment (BaihuiJin) [reply optimum/habana/transformers/models/gpt2/modeling_gpt2.py:78]: Yes, it still has perf drop if we remove this.

File: optimum/habana/transformers/models/gpt2/modeling_gpt2.py (hunk 2 above)
Comment (BaihuiJin) [reply optimum/habana/transformers/models/gpt2/modeling_gpt2.py:120]: Yes, branch has been tested.

**Review by ZhaiFeiyue — APPROVED** [review]
LGTM

**Comment by ZhaiFeiyue** [comment]
@regisss let's merge this PR?

**Pull request merged**